gnmidiff/setrequest.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2023 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// gnmidiff contains gNMI utilities for diffing SetRequests and GetResponses.
+package gnmidiff
+
+import (
+	"fmt"
+
+	"github.com/derekparker/trie"
+	"github.com/openconfig/ygot/ygot"
+	"github.com/openconfig/ygot/ytypes"
+
+	gpb "github.com/openconfig/gnmi/proto/gnmi"
+)
+
+// MismatchedUpdate represents two different update values for the same leaf
+// node.
+type MismatchedUpdate struct {
+	// A is the update value in A.
+	A interface{}
+	// B is the update value in B.
+	B interface{}
+}
+
+// SetRequestDiff contains the intent difference between two SetRequests.
+//
+// - The key of the maps is the string representation of a gpb.Path constructed
+// by ygot.PathToString.
+// - The value of the update fields is the JSON_IETF representation of the
+// value.
+type SetRequestDiff struct {
+	AOnlyDeletes      map[string]struct{}
+	BOnlyDeletes      map[string]struct{}
+	CommonDeletes     map[string]struct{}
+	AOnlyUpdates      map[string]interface{}
+	BOnlyUpdates      map[string]interface{}
+	CommonUpdates     map[string]interface{}
```

Comment:
Is it ever of interest to a caller whether updated values were as a result of a replace or an update operation in the `SetRequest`? I couldn't immediately think of a case where it would be, but wanted to check.

Reply:
This makes sense. I can add a feature where these updates and deletes are each associated with the `Update` message in the SetRequest. If there is a need then we can add it.

Reply:
Yes. In that case the delete will be stored in the `delete` field of `setRequestIntent`. The deletions in the `setRequestIntent` always precede the `update` field. So it's basically reducing the initial SetRequest down to just deletes and updates.